crates/bevy_ptr/src/lib.rs

```diff
@@ -285,6 +313,183 @@ impl_ptr!(Ptr);
+        mem::forget(self);
+        value
+    }
+}
+
+impl<'a, T, A: IsAligned> MovingPtr<'_, T, A> {
+    /// Creates a new instance from a raw pointer.
+    ///
+    /// # Safety
+    /// - `inner` must point to valid value of `T`.
+    /// - If the `A` type parameter is [`Aligned`] then `inner` must be sufficiently aligned for `T`.
+    /// - `inner` must have correct provenance to allow read and writes of the pointee type.
+    /// - The lifetime `'a` must be constrained such that this [`MovingPtr`] will stay valid and nothing
+    ///   else can read or mutate the pointee while this [`MovingPtr`] is live.
+    #[inline]
+    pub unsafe fn new(inner: NonNull<T>) -> Self {
+        Self(inner, PhantomData)
+    }
+
+    /// This exists mostly to reduce compile times;
+    /// code is only duplicated per type, rather than per function called.
+    fn make_internal(temp: &mut ManuallyDrop<T>) -> MovingPtr<'_, T> {
+        MovingPtr(NonNull::from(temp).cast::<T>(), PhantomData)
+    }
+
+    /// Consumes a value and creates an [`MovingPtr`] to it while ensuring a double drop does not happen.
+    #[inline]
+    pub fn make<F: FnOnce(MovingPtr<'_, T>) -> R, R>(val: T, f: F) -> R {
+        let mut val = ManuallyDrop::new(val);
+        // SAFETY: The value behind the pointer will not get dropped or observed later,
+        // so it's safe to promote it to an owning pointer.
+        f(Self::make_internal(&mut val))
+    }
+
+    /// Partially moves out some fields inside of `self`.
+    ///
+    /// The partially returned value is returned back pointing to `MaybeUninit<T>`.
+    ///
+    /// # Safety
+    /// The fields moved out of in `f` must not be accessed or dropped after this function returns.
+    #[inline]
+    pub unsafe fn partial_move(
+        ptr: MovingPtr<'a, T, A>,
+        f: impl FnOnce(MovingPtr<'a, T, A>) -> MovingPtr<'a, T, A>,
+    ) -> MovingPtr<'a, MaybeUninit<T>, A> {
+        let value = f(ptr);
+        let partial = MovingPtr(value.0.cast::<MaybeUninit<T>>(), PhantomData);
+        mem::forget(value);
+        partial
+    }
+
+    /// Reads the value pointed to by this pointer.
+    #[inline]
+    pub fn read(self) -> T {
+        let value = if const { A::IS_ALIGNED } {
+            unsafe { self.0.as_ptr().read() }
+        } else {
+            unsafe { self.0.as_ptr().read_unaligned() }
+        };
+        mem::forget(self);
+        value
+    }
+
+    /// Writes the value pointed to by this pointer to a provided location.
+    ///
+    /// This does *not* drop the value stored at `dst` and it's the caller's responsibility
+    /// to ensure that it's properly dropped.
+    ///
+    /// # Safety
+    ///  - `dst` must be valid for writes.
+    ///  - If the `A` type parameter is [`Aligned`] then `dst` must be properly aligned for `T`.
+    #[inline]
+    pub unsafe fn write_to(self, dst: *mut T) {
+        let src = self.0.as_ptr();
+        mem::forget(self);
+        if const { A::IS_ALIGNED } {
```

Comment:
Learning question: what does the use of `const` in this position do? Where can I read more about it?

Reply:
I went ahead and turned these into associated functions instead.